Design/Instrument-2.0/ScanningInstruments_Design.md

```diff
@@ -0,0 +1,102 @@
+
+### Design
+
+#### `DetectorInfo`
+
+The core of the changes to support scanning instruments will be in the [`DetectorInfo`](https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/DetectorInfo.h) class. There are two choices for implementing the required changes in `DetectorInfo`. Currently the methods `isMonitor`, `isMasked`, `l2`, `twoTheta`, `signedTwoTheta`, `position`, `rotation`, `setPosition` and `setRotationEach` are accessed by a detector index. This could be kept and the detector index would be associated with a detector ID and a time index. Alternatively the methods can be accessed by supplying the detector index, which has a one-to-one correspondence with a detector ID, and a time index. There would be a time index corresponding to each position in the step scan.
+
+The current plan is to access the methods using the detector index and a time index. This will save memory on detector metadata which would not change for different time indices, such as the monitor flag. See the end of this document for the changes to the signature of some of the public `DetectorInfo` methods.
+
+The `timeIndex` argument is optional, so algorithms that do not care about time indexing will simply return the first/only time index. `DetectorInfo` should also contain a vector of start and end times for each time index. This will mean any information stored as a time series sample log can also be obtained for the correct step, for example the monitor counts for normalisation.
+
+`DetectorInfo` will ultimately store the information for positions and rotation, so will need to do this in a way that can be accessed by both the detector index and time index. Masking could also be stored in a time indexed fashion.
+
+#### `SpectrumInfo`
+
+Currently [`SpectrumInfo`](https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/SpectrumInfo.h) has a `getDetector` method for finding the detector for the workspace. This uses the `getSpectrum(index)` method on `MatrixWorkspace` to return the detector IDs for the spectrum.
+
+A new method should be added, for example `getTimeIndexedDetectorIDs(index)` to return a set of `std::pair<detid_t, size_t>`.
+
+This can be used to access methods on `DetectorInfo`, for example `m_detectorInfo.twoTheta(detIndex, timeIndex)`.
+
+#### `MatrixWorkspace` & `ISpectrum`/`Spectrum`
+
+As metioned under **`SpectrumInfo`** `MatrixWorkspace` will require a `getTimeIndexedDetectorIDs(index)` method.
+
+This implies the member variable `std::set<detid_t> detectorIDs` in `ISpectrum` will need to become `std::set<std::pair<detid_t, size_t>> timeIndexedDetectorIDs`.
```

Comment:
I think you are probably right that this might be needed as a temporary workaround. Where possible, we should however try to eliminate the need to directly work with detector IDs. Once we have a fully functional `SpectrumInfo` and `DetectorInfo` there should be very few cases where algorithms need to detector IDs.